### PR TITLE
Fix race and speed up amp-list/amp-bind interaction

### LINKED ISSUE #22341: Slow Rendering for amp-list when binding="always" enabled
## What's the issue?

There is a significant delay when an amp-list has binding="always" enabled. The delay is exaggerated on production pages when more amp-components and other elements are present on the page.

Screen Shot 2019-05-16 at 11 54 18 AM

Screen Shot 2019-05-16 at 11 54 30 AM

Video of Delay
https://screencast-o-matic.com/watch/cqhDqzTecy

## How do we reproduce the issue?

https://wompmobileamp.azureedge.net/7605/ampBindListSlow.html

1. Visit the page above in Chrome Mobile Emulator with these settings
a. 9Mbps network speed
b. 4x CPU reduction

## What browsers are affected?

Chrome on slower Android devices
Chrome Mobile Emulator with 4x CPU reduction

## Which AMP version is affected?

Current version

### PR DESCRIPTION
**Pull request by choumx**
Fixes #22341, partial for #21901.

- Bug: Fix race condition between `amp-list[binding=refresh]` and amp-bind's initial DOM scan.
- Optimization: Skip `Bind.rescan` in amp-list if new children have no binding marker attributes (`[i-amphtml-binding]`).
- Optimization: Significantly speed up initial render for amp-list[binding=always] by changing `Bind.rescan` to only wait for a small part of Bind's initialization.
- Bug: Refactor `Bind.rescan` to fix/support amp-date-picker's usage.

### CODE REVIEW

File: extensions/amp-date-picker/0.1/amp-date-picker.js [hunk 1]
```diff
@@ -1876,10 +1876,10 @@ export class AmpDatePicker extends AMP.BaseElement {
   onMount() {
     if (this.mode_ == DatePickerMode.OVERLAY) {
       // REVIEW: this should be ok, since opening the overlay requires a
-      // user interaction, and this won't run until then
+      // user interaction, and this won't run until then.
       Services.bindForDocOrNull(this.element).then(bind => {
         if (bind) {
-          return bind.scanAndApply([this.element], [this.element]);
+          return bind.rescan([this.element], [this.element], {'apply': true});
```
Comment (choumx) [inline extensions/amp-date-picker/0.1/amp-date-picker.js]: @cvializ I think this was broken for awhile. 🤷‍♂

Comment (cvializ) [reply extensions/amp-date-picker/0.1/amp-date-picker.js]: 😬

File: extensions/amp-bind/0.1/bind-impl.js [hunk 2]
```diff
@@ -394,92 +405,105 @@ export class Bind {
...
+    // Sync remove bindings from internal state first, but don't chain on
+    // returned promise (worker message) as an optimization.
+    const removePromise = this.removeBindingsForNodes_(removedElements);
+
+    // Scan `addedElements` and descendants for bindings.
     const bindings = [];
-    // Scan `addedElements` and their children for elements with bindings.
-    const elementsToScan = addedElements.slice();
+    const elementsToScan = addedElements.filter(el =>
+      el.hasAttribute('i-amphtml-binding')
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-impl.js]: Why are you filtering here?

Comment (choumx) [reply extensions/amp-bind/0.1/bind-impl.js]: Minor optimization. We only want to scan elements with `[i-amphtml-binding]` attribute, and elements in `addedElements` may not have it. 

`scanElement_()` iterates over an element's attributes, so this should be a small perf win.

**Pull request merged**